olive/model/hf/hf_model.py

```diff
@@ -0,0 +1,98 @@
+    EncoderInit = "encoder_init"
+    DecoderInit = "decoder_init"
+    # Will be deprecated
+    EncoderDecoderInit = "encoder_decoder_init"
+
+
+class HFComponent(ConfigBase):
+    type: HFComponentType
+    io_config: Optional[IOConfig] = None
+    component_func: Optional[Union[str, Callable]] = None
+    dummy_inputs_func: Optional[Union[str, Callable]] = None
+
+
+class HFConfig(ConfigBase):
+    model_name: str = None
+    task: str = None
+    # TODO: remove model_class and only use task
+    model_class: str = None
+    model_type: str = None
+    state_dict_path: Union[str, Path] = None
+    use_ort_implementation: bool = False
+    use_custom_implementation: bool = False
+    components: List[HFComponent] = None
+    config: Dict[str, Any] = None
+    dataset: Dict[str, Any] = None
+
+    @validator("model_class", always=True)
+    def task_or_model_class_required(cls, v, values):
+        if values["model_name"]:
+            if not v and not values.get("task", None):
+                raise ValueError("Either task or model_class must be specified")
+            return v
+
+    @validator("components")
+    def validate_components(cls, v, values):
+        if values["use_custom_implementation"]:
+            component_set = {component.type for component in v}
+            if "encoder" not in component_set and "decoder" not in component_set:
+                raise ValueError("Either encoder or decoder must be specified")
+            if "encoder_init" in component_set and "encoder" not in component_set:
+                raise ValueError("encoder_init can only be specified when encoder is specified")
+            if "decoder_init" in component_set and "decoder" not in component_set:
+                raise ValueError("decoder_init can only be specified when decoder is specified")
+            for component in v:
+                if component.io_config is None:
+                    raise ValueError("io_config must be specified for all components")
+                if component.component_func is None:
+                    raise ValueError("component_func must be specified for all components")
+                if component.dummy_inputs_func is None:
+                    raise ValueError("dummy_inputs_func must be specified for all components")
+        return v
+
+
+class HFModelBase(torch.nn.Module):
+    """Base class for all HF models"""
+
+    def __init__(
+        self,
+        encoder: Optional[torch.nn.Module] = None,
+        decoder: Optional[torch.nn.Module] = None,
+        encoder_init: Optional[torch.nn.Module] = None,
+        decoder_init: Optional[torch.nn.Module] = None,
+        config: Optional[Dict[str, Any]] = None,
+    ):
+        super().__init__()
+        self.encoder = encoder
+        self.decoder = decoder
+        self.encoder_init = encoder_init
+        self.decoder_init = decoder_init
+        self.config = config
+
+
+class ORTWhisperModel(torch.nn.Module):
+    """ORT implementation of whisper model"""
```

Comment:
This should be outside the Olive, in the customer code.

Reply:
I'll remove this once I refactor whisper example.